### PR TITLE
Add support for baremetal platform

### PR DESCRIPTION
**Pull request by stbenjam**
For KNI, we need to support the baremetal platform in MCO.

### CODE REVIEW

File: templates/master/01-master-kubelet/baremetal/files/cloudconfig.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+filesystem: "root"
```
Comment (abhinavdahiya) [inline templates/master/01-master-kubelet/baremetal/files/cloudconfig.yaml]: Each file doesn't need to be added, only special platform override.

File: templates/master/01-master-kubelet/baremetal/files/kubelet.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+filesystem: "root"
```
Comment (abhinavdahiya) [inline templates/master/01-master-kubelet/baremetal/files/kubelet.yaml]: This also looks like the default one. This shouldn't be required.

File: templates/master/01-master-kubelet/baremetal/units/kubelet.yaml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  ExecStart=/usr/bin/hyperkube \
+      kubelet \
+        --config=/etc/kubernetes/kubelet.conf \
+        --bootstrap-kubeconfig=/etc/kubernetes/kubeconfig \
+        --rotate-certificates \
+        --kubeconfig=/var/lib/kubelet/kubeconfig \
+        --container-runtime=remote \
+        --container-runtime-endpoint=/var/run/crio/crio.sock \
+        --allow-privileged \
+        --node-labels=node-role.kubernetes.io/master,node.openshift.io/os_version=${VERSION_ID},node.openshift.io/os_id=${ID} \
```
Comment (abhinavdahiya) [inline templates/master/01-master-kubelet/baremetal/units/kubelet.yaml]: personal opinion only, feel free to ignore.

Do you have a strong reason why this platform is deviating from other platforms of OpenShift.

This deviation is not the best way forward. It adds another dimension of things operators need to know for this platform as implicit API

A better solution would be to teach operators to explicitly tolerate that taint for this platform.

File: templates/master/01-master-kubelet/baremetal/files/kubelet.yaml (hunk 2 above)
Comment (stbenjam) [reply templates/master/01-master-kubelet/baremetal/files/kubelet.yaml]: This may need changing - we have some questions about how the max pods was defined:  https://github.com/openshift/machine-config-operator/issues/723

File: templates/master/01-master-kubelet/baremetal/files/cloudconfig.yaml (hunk 1 above)
Comment (stbenjam) [reply templates/master/01-master-kubelet/baremetal/files/cloudconfig.yaml]: Thanks, I didn't realize - removed

File: templates/master/01-master-kubelet/baremetal/units/kubelet.yaml (hunk 3 above)
Comment (stbenjam) [reply templates/master/01-master-kubelet/baremetal/units/kubelet.yaml]: > Do you have a strong reason why this platform is deviating from other platforms of OpenShift.

For KNI, the hardware has a starting point of around 40 cores and 256GB of ram each. It's a big expense to dedicate that much hardware only to the control plane.  A larger cluster could be willing to have that much overhead, but our minimum is 4 servers, which means having 75% overhead.

Could we make the NoSchedule taint configurable through the installer for all platforms, and avoid the special case for baremetal?

> A better solution would be to teach operators to explicitly tolerate that taint for this platform.

How many operators would that be? It seems preferable to me to do it in one place, rather than many. Also: would users need to be aware of this if they wanted to run their own workloads on the baremetal platform? What if they're using operators we don't control?

Comment (russellb) [reply templates/master/01-master-kubelet/baremetal/units/kubelet.yaml]: I actually don't think this should be the default for the baremetal platform.  We should default to the same as any other platform.

What's needed is a way to declaratively specify that workloads should be able to run on masters.  We have a product under development that would use this by default, but that's different from the more general IPI baremetal support.  There's a 4.2 epic for this "workloads on masters" support.

In short, this part should probably be dropped for now, and wait for the fully supported way to specify this.

**Comment by DanyC97** [comment]
@stbenjam you mind asking what KNI stands for ?

**Comment by russellb** [comment]
> @stbenjam you mind asking what KNI stands for ?

Kubernetes Native Infrastructure.

Part of that is IPI bare metal support

**Comment by stbenjam** [comment]
I removed the customizations, and generated the test data.

**Comment by cgwalters** [comment]
A few PRs landed in master that clean up the template stuff, I think you want to rebase on those;

https://github.com/openshift/machine-config-operator/pull/749
https://github.com/openshift/machine-config-operator/pull/653

> This includes a custom kubelet configuration that does not have the NoSchedule taint on
masters.

Is that the main effect of this change?  Personally...I feel like this should be an installConfig flag or something rather than part of the platform.   This is [explicitly documented in the installer docs](https://github.com/openshift/installer/blob/0f5ab04c23339842d58ac6a1ad1923954197b997/docs/user/customization.md#control-plane-with-no-taints) today, it's just an ergonomic hit to use.  If we have enough use cases for it I think it's worth raising up to be an explicit flag.

And hmm oh god that explicitly documents overriding an already extant file which is going to be painful for Ignition spec3 migration.  Further we should really make this a drop-in or something rather than requiring people to duplicate the `kubelet.service`.

**Comment by stbenjam** [comment]
> A few PRs landed in master that clean up the template stuff, I think you want to rebase on those;
> 
> #749
> #653

Thanks, will rebase

> > This includes a custom kubelet configuration that does not have the NoSchedule taint on
> > masters.
> 
> Is that the main effect of this change? Personally...I feel like this should be an installConfig flag or something rather than part of the platform. This is [explicitly documented in the installer docs](https://github.com/openshift/installer/blob/0f5ab04c23339842d58ac6a1ad1923954197b997/docs/user/customization.md#control-plane-with-no-taints) today, it's just an ergonomic hit to use. If we have enough use cases for it I think it's worth raising up to be an explicit flag.

Well, that was originally the point of this change - but we've decided against it for now, and are looking for another mechanism. See https://github.com/openshift/machine-config-operator/pull/722#discussion_r282350644

I updated this PR anyway, as it probably makes sense to have the baremetal platform here so it's like all the others, right? We are working towards eliminating our installer fork, and have incorporated the baremetal platform into other parts already like cluster/api, but not MCO explicitly.

> And hmm oh god that explicitly documents overriding an already extant file which is going to be painful for Ignition spec3 migration. Further we should really make this a drop-in or something rather than requiring people to duplicate the `kubelet.service`.

File: templates/master/01-master-kubelet/baremetal/units/kubelet.yaml (hunk 3 above)
Comment (cgwalters) [reply templates/master/01-master-kubelet/baremetal/units/kubelet.yaml]: I very much agree with https://github.com/openshift/machine-config-operator/pull/722#discussion_r283352069 - we want a `controlPlane: schedulable` bool or something.
See https://github.com/openshift/machine-config-operator/pull/722#issuecomment-492828738 and following.

**Comment by cgwalters** [comment]
Moved the master schedulable thing to https://github.com/openshift/machine-config-operator/issues/763

**Comment by stbenjam** [comment]
https://github.com/openshift/machine-config-operator/pull/795 adds the baremetal platform. Don’t need this one.

**Comment by stbenjam** [comment]
Re-opening this, we'll need it temporarily.

Currently, for baremetal IPI and our installer fork, we have to bail out of the installer early to do some things - one of those things is disabling NoSchedule taints: https://github.com/openshift-metal3/dev-scripts/blob/master/06_create_cluster.sh#L82-L83

We are anxious to get an experimental version of this platform into openshift/installer. This adds the barematel platform to MCO, and overrides the kubelet config.  Once #763 is resolved, we can move to the new mechanism, or alternatively once we have workers being deployed in the installer we won't need the override at all (this is currently a work in progress).

**Comment by DanyC97** [comment]
@stbenjam what is the history behind the need to disable NoSchedule taint ? just curious, thanks
